### PR TITLE
Early return in `ungroup()`

### LINKED ISSUE #684: `ungroup()` should not add operation on ungrouped tbl
``` r
library(dplyr, warn.conflicts = FALSE)
library(dbplyr)
#> 
#> Attaching package: 'dbplyr'
#> The following objects are masked from 'package:dplyr':
#> 
#>     ident, sql

lazy_frame(x = 1) %>% 
  ungroup() %>% 
  str()
#> List of 2
#>  $ ops:List of 4
#>   ..$ name: chr "ungroup"
#>   ..$ x   :List of 2
#>   .. ..$ x   : tibble [1 × 1] (S3: tbl_df/tbl/data.frame)
#>   .. .. ..$ x: num 1
#>   .. ..$ vars: chr "x"
#>   .. ..- attr(*, "class")= chr [1:3] "op_base_local" "op_base" "op"
#>   ..$ dots: list()
#>   ..$ args: list()
#>   ..- attr(*, "class")= chr [1:3] "op_ungroup" "op_single" "op"
#>  $ src:List of 2
#>   ..$ con  : list()
#>   .. ..- attr(*, "class")= chr [1:2] "TestConnection" "DBIConnection"
#>   ..$ disco: NULL
#>   ..- attr(*, "class")= chr [1:4] "src_TestConnection" "src_dbi" "src_sql" "src"
#>  - attr(*, "class")= chr [1:3] "tbl_TestConnection" "tbl_lazy" "tbl"
```

<sup>Created on 2021-08-02 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

### LINKED ISSUE #689: `ungroup(lf, x)` should only remove `x` from group
Instead of only removing `x` from the groups it removes all grouping

``` r
library(dplyr, warn.conflicts = FALSE)
library(dbplyr)
#> 
#> Attaching package: 'dbplyr'
#> The following objects are masked from 'package:dplyr':
#> 
#>     ident, sql

lazy_frame(x = 1, y = 2) %>% 
  group_by(x, y) %>% 
  ungroup(x) %>% 
  group_vars()
#> character(0)
```

<sup>Created on 2021-08-16 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

### PR DESCRIPTION
**Pull request by mgirlich**
Fixes #684 and fixes #689.

* remove ungroup operation and let `group_by()` do the work
* `group_by()` and `ungroup()` don't add an operation if nothing has changed
* `ungroup(df, x)` now works
* remove check if `.add` is in formals of `group_by()` as `dbplyr` depends on `dplyr >= 1.0.4` and `.add` was added with 1.0.0

### CODE REVIEW

File: R/verb-group_by.R [hunk 1]
```diff
@@ -37,20 +37,12 @@ group_by.tbl_lazy <- function(.data, ..., .add = FALSE, add = NULL, .drop = TRUE
...
-    }
-  }
+  groups <- dplyr::group_by_prepare(.data, !!!dots, .add = .add)
+  names <- purrr::map_chr(groups$groups, as_string)
 
-  if (".add" %in% names(formals("group_by"))) {
-    groups <- dplyr::group_by_prepare(.data, !!!dots, .add = .add)
-  } else {
-    groups <- dplyr::group_by_prepare(.data, !!!dots, add = .add)
+  if (can_group_by_can_return_early(.data, groups$data, groups$groups)) {
```
Comment (krlmlr) [inline R/verb-group_by.R:43]: Should we inline this function? I'm not sure how to read `setequal(.data, ...)` -- should we do `setequal(group_vars(.data), ...)` or the like?

File: tests/testthat/test-verb-group_by.R [hunk 2]
```diff
@@ -48,9 +48,36 @@ test_that("group_by handles empty dots", {
...
+
+  out <- lf_grouped %>% mutate(y = y - mean(y)) %>% group_by()
+  expect_s3_class(out$ops, "op_group_by")
+  expect_equal(group_vars(out), character())
+})
+
+test_that("group_by mutate is not optimised away", {
+  lf <- lazy_frame(x = 1) %>% group_by(x)
+
+  expect_equal(
```
Comment (krlmlr) [inline tests/testthat/test-verb-group_by.R:75]: Should this be a snapshot test?

Comment (mgirlich) [reply tests/testthat/test-verb-group_by.R:75]: I don't think so: the expectation is very simple, I know it explicitly, and I don't expect this to change.

File: R/verb-group_by.R (hunk 1 above)
Comment (mgirlich) [reply R/verb-group_by.R]: I realised that this can be done much simpler anyway 😄

File: R/verb-group_by.R [hunk 3]
```diff
@@ -37,20 +37,13 @@ group_by.tbl_lazy <- function(.data, ..., .add = FALSE, add = NULL, .drop = TRUE
...
-    }
-  }
+  groups <- dplyr::group_by_prepare(.data, !!!dots, .add = .add)
+  names <- purrr::map_chr(groups$groups, as_string)
 
-  if (".add" %in% names(formals("group_by"))) {
-    groups <- dplyr::group_by_prepare(.data, !!!dots, .add = .add)
-  } else {
-    groups <- dplyr::group_by_prepare(.data, !!!dots, add = .add)
+  group_by_can_return_early <- setequal(groups$group_names, group_vars(.data))
```
Comment (hadley) [inline R/verb-group_by.R:43]: Maybe call this `same_groups`?

**Review by hadley — APPROVED** [review]

**Pull request merged**